---
title: "Curso dplyr y ggplot2"
author: "BioInfoGRX - RSG Spain"
date: "`r format(Sys.Date(), '%d/%m/%Y')`"
output:
  html_document:
    df_print: default
    code_folding: show
    number_sections: yes
    theme: lumen
    toc: yes
    toc_depth: 3
    toc_float:
      collapsed: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introducción

Paquetes necesarios:

```{r, message=FALSE, warning=FALSE, error=FALSE}
# TO-DO: Indicar brevemente para qué es cada paquete
library(dplyr)
library(ggplot2)
library(DT)
library(pasilla) #Bioconductor
library(reshape2)
library(tidyr) ## pivot functions
library(tibble)
```


# Ejemplos de juguete

El formato ideal de tabla para representar datos de expresión génica con ggplot2 requiere una fila para cada valor de expresión:

```{r}
gene_expression = data.frame(gene = rep(paste0("Gene",seq(1,10)), c(rep(4,10))),
                             sample = rep(paste0("Sample",seq(1,4)),10),
                             group = rep(paste0("Group",seq(1,2)),20),
                             value = rnorm(40))

datatable(gene_expression)
```

```{r, fig.align='center'}
a = ggplot(data = gene_expression,aes(x=sample,y=value))+
  geom_boxplot()
print(a)
ggsave("Figures/1_toyexample.png",plot = a)
```


# Paquete `dplyr`

## Operaciones básicas

```{r}
##### Basic Operation #####

data(starwars)
datatable(starwars)
```


```{r}
## select: subset column ##
select(starwars,name,height,mass,species,homeworld)
select(starwars,name:homeworld)
```


```{r}
## filter: subset rows by condition ##
filter(starwars,species=="Droid")
filter(starwars,species=="Human",mass < 100)
filter(starwars,species=="Human" | mass < 100)
```


```{r}
## mutate: create a new column
mutate(starwars,height_by_mass = height / mass)
```


```{r}
## group_by: group data
group_by(starwars,gender)
```


```{r}
## summarize: create stats from group data
summarize(group_by(starwars,gender),mean_mass = mean(mass,na.rm=T))
summarize(group_by(starwars,gender),mean_height = mean(height,na.rm=T))
```


```{r}
## arrange: sort columns
arrange(starwars,mass)
arrange(starwars,desc(mass))
```


```{r}
## count: count discrete values
count(starwars,sex)
```


```{r}
## pipes: the usual way to use dplyr --> dataframe %>% operation1 %>% operation 2 .....
starwars %>% select(name,height)
starwars %>% group_by(gender) %>% summarise(mean_mas=mean(mass,na.rm=T))
starwars %>% select(name:species) %>% filter(species != "Human") %>% group_by(sex) %>% summarize(max_height = max(height,na.rm=T))
```


## Combinando dataframes

```{r}
superheroes = data.frame(name = c("Magneto","Storm","Mystique","Batman","Joker","Catwoman","Hellboy"),
                         alignment = c("bad","good","bad","good","bad","bad","good"),
                         gender = c("male","female","female","male","male","female","male"),
                         publisher = c("Marvel","Marvel","Marvel","DC","DC","DC","Dark Horse Comics"))

publishers = data.frame(publisher = c("DC","Marvel","Image"),
                        year_founded = c(1934,1939,1992))
```


```{r}
## Inner join: MATCH IN BOTH TABLES
superheroes %>% inner_join(publishers,by="publisher")
```


```{r}
## Left join: prioritizes left table
superheroes %>% left_join(publishers,by="publisher")
```


```{r}
## Right join: prioritizes right table
publishers %>% right_join(superheroes,by="publisher")

## Full join: retains all rows
superheroes %>% full_join(publishers,by="publisher")
```


```{r}
## Semi join: match in both tables but do not add the information from the second table
superheroes %>% inner_join(publishers,by="publisher")
superheroes %>% semi_join(publishers,by="publisher")
```


```{r}
## anti_join: contrary of semi_join
superheroes %>% anti_join(publishers,by="publisher")
```


```{r}
superheroes %>% inner_join(publishers,by="publisher") %>% filter(alignment == "good")
```


## Preparando datos para ggplot2

```{r}
gene_expression = data.frame(gene = rep(paste0("Gene",seq(1,10)), c(rep(4,10))),
                             sample = rep(paste0("Sample",seq(1,4)),10),
                             group = rep(paste0("Group",seq(1,2)),20),
                             values = rnorm(40))


pasCts <- system.file("extdata",
                      "pasilla_gene_counts.tsv",
                      package="pasilla", mustWork=TRUE)
pasAnno <- system.file("extdata",
                       "pasilla_sample_annotation.csv",
                       package="pasilla", mustWork=TRUE)
cts <- as.matrix(read.csv(pasCts,sep="\t",row.names="gene_id"))
datatable(cts)
```


```{r}
coldata <- read.csv(pasAnno, row.names=1)
coldata <- coldata[,c("condition","type")]
rownames(coldata) = colnames(cts)
datatable(coldata)
```


```{r}
######## The logical way --> Loop ##########
gene_expression = data.frame(matrix(NA,ncol = 3))[0,]
colnames(gene_expression) = c("sample","gene","value")

for (samplename in colnames(cts)){
  gene_vals = cts[,samplename]
  gene_expression = rbind(gene_expression,data.frame(sample = samplename,gene = rownames(cts),value = gene_vals))
}

coldata$sample = rownames(coldata)
gene_expression = merge(gene_expression,coldata,by="sample")
```


```{r}
######## Using Tibble and Dplyr or Reshape2 and Dplyr ##########
gene_expression = as.data.frame(cts) %>% 
  rownames_to_column(var = "gene") %>%  
  pivot_longer(cols = rownames(coldata),names_to = "sample",values_to = "value") %>% 
  full_join(coldata,by="sample")

gene_expression = melt(cts, varnames=c('gene', 'sample')) %>% full_join(coldata,by="sample")
```


```{r}
### Use this gene expression in toy_example 2

######## gene_expression to cts

cts = gene_expression[,c("gene","sample","value")] %>% pivot_wider(names_from = "sample", values_from = "value") %>%
  column_to_rownames(var = "gene") %>% as.matrix()

cts = dcast(gene_expression, gene ~ sample, value.var = "value") %>%
  column_to_rownames(var = "gene") %>% as.matrix()


```

```{r}
a = ggplot(data = gene_expression,aes(x=sample,y=value))+
  geom_boxplot()
print(a)
ggsave("Figures/2_toyexample.png",plot = a)
```


```{r}

```

# Paquete `ggplot2` (TO-DO)

# Session info

```{r}
sessionInfo()
```


